### PR TITLE
Renesas R-Car Add "Interrupt driven UART"

### PR DESCRIPTION
**Pull request by aaillet**
This PR add the support of "Interrupt driven UART"
to the R-Car UART driver and enable it for the H3ULCB board.
    
The driver is supporting nearly all the methods that are
described in the "Interrupt driven UART" part of the
uart_driver_api.
    
This new version of the driver has been tested on
H3ULCB board by running "uart_basic_api" test suite.

### CODE REVIEW

**Review by julien-massot — APPROVED** [review]

**Comment by aaillet** [comment]
Fixed reset of IRQ flag for each byte instead of waiting
for the whole payload to be read/write before
resetting the flag.

FIFO read was broken and was receiving the same byte
infinitely because the flag wasn't reset.

**Comment by dcpleung** [comment]
It would be better if you add a spinlock to make sure all the read-modify-write of registers are atomic.

**Comment by aaillet** [comment]
Just added a spinlock to the driver, please @dcpleung tell me if it's better like that. Thanks

File: drivers/serial/uart_rcar.c [hunk 1]
```diff
@@ -258,33 +289,312 @@ static int uart_rcar_init(const struct device *dev)
...
+{
+	const struct uart_rcar_cfg *config = DEV_UART_CFG(dev);
+	struct uart_rcar_data *data = DEV_UART_DATA(dev);
+
+	uint16_t reg_val;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	reg_val = !!(uart_rcar_read_16(config, SCSCR) & irq);
+
+	k_spin_unlock(&data->lock, key);
```
Comment (dcpleung) [inline drivers/serial/uart_rcar.c:317]: If the hardware can read the register atomically, there should be no need to wrap it around a spin lock. I am not familiar with the target hardware. So if you think this locking is needed, I am okay with it. (ditto for two more read-only locking below)

**Comment by aaillet** [comment]
Thank you @dcpleung for your comment, after checking, the hardware used can read and write registers atomically (when there is no need to read/write/modify). So, I just removed some useless locking !

**Review by dcpleung — APPROVED** [review]

**Pull request merged**